corehq/apps/translations/views.py

```diff
@@ -0,0 +1,195 @@
+
+from corehq.apps.hqwebapp.decorators import use_select2
+from corehq.apps.translations.forms import (
+    ConvertTranslationsForm,
+    PullResourceForm,
+)
+from corehq.apps.app_manager.app_translations.generators import Translation, PoFileGenerator
+from corehq.util.files import safe_filename_header
+from corehq.apps.domain.views import BaseDomainView
+from custom.icds.translations.integrations.exceptions import ResourceMissing
+from custom.icds.translations.integrations.transifex import Transifex
+from custom.icds.translations.integrations.utils import get_file_content_from_workbook
+
+
+class ConvertTranslations(BaseDomainView):
+    page_title = _('Convert Translations')
+    urlname = 'convert_translations'
+    template_name = 'convert_translations.html'
+    section_name = ugettext_noop("Translations")
+
+    @property
+    @memoized
+    def convert_translation_form(self):
+        if self.request.POST:
+            return ConvertTranslationsForm(self.request.POST, self.request.FILES)
+        else:
+            return ConvertTranslationsForm()
+
+    @property
+    @memoized
+    def _uploaded_file_name(self):
+        uploaded_file = self.convert_translation_form.cleaned_data.get('upload_file')
+        return uploaded_file.name
+
+    @staticmethod
+    def _parse_excel_sheet(worksheet):
+        """
+        :return: the rows and the index for expected columns
+        """
+        rows = [row for row in worksheet.iter_rows()]
+        headers = [cell.value for cell in rows[0]]
+        source = headers.index('source')
+        translation = headers.index('translation')
+        occurrence = headers.index('occurrence') if 'occurrence' in headers else None
+        context = headers.index('context') if 'context' in headers else None
+        return rows, source, translation, occurrence, context
+
+    def _generate_translations_for_po(self, worksheet):
+        """
+        :return: a list of Translation objects
+        """
+        rows, source, translation, occurrence, context = self._parse_excel_sheet(worksheet)
+        translations = {worksheet.title: []}
+        for index, row in enumerate(rows[1:]):
+            _occurrence = row[occurrence].value if occurrence is not None else ''
+            _context = row[context].value if context is not None else ''
+            translations[worksheet.title].append(
+                Translation(
+                    row[source].value,
+                    row[translation].value,
+                    [(_occurrence, None)],
+                    _context)
+            )
+        return translations
+
+    def _generate_po_file(self, worksheet):
+        """
+        extract translations from worksheet and converts to a po file
+        :return: PoFileGenerator object
+        """
+        translations = self._generate_translations_for_po(worksheet)
+        po_file_generator = PoFileGenerator()
+        po_file_generator.generate_translation_files(translations, {})
```

Comment:
Is there a reason not to generate translations as part of the constructor? This could be shortened to
```py
po_file_generator = PoFileGenerator(translations, {})
```
As is, `PoFileGenerator` is in an invalid state on initial construction until `generate_translation_files` is called. This is an OO anti-pattern. Some things that might invalidate this suggestion:

- if `generate_translation_files` is called more than once for a given `PoFileGenerator` instance.
- if a `PoFileGenerator` instance is ever used without calling it's `generate_translation_files` method.

Reply:
Does [this](https://github.com/dimagi/commcare-hq/blob/425d72efd19083a79e37c34643fc5d002528af63/corehq/apps/app_manager/app_translations/generators.py#L49) look like a valid invalidation? :)

Reply:
Not necessarily. I can't tell if `self.po_file_generator` is referenced anywhere before `self.po_file_generator.generate_translation_files()` is called, and if it is, is its state (not having generated translation files yet) intentional at that point, or is it a bug to have referenced it before translations have been generated?

If it is not referenced before `generate_translation_files()` is called, then why does `self.po_file_generator` need to be created/assigned beforehand?

Reply:
`po_file_generator` is used at [other places](https://github.com/dimagi/commcare-hq/blob/0a04cd5ffa39bc2a0e5825bdf49d46c3a5bbac6e/corehq/apps/app_manager/app_translations/generators.py#L191-L200) so just wanted to keep a reference on the object.
Also personally don't like `__init__` doing something major. I like `__init__` to just initialise things and be simple. But if its not with the OO pattern or not the python/django way of doing things, happy to modify it like you suggested.

Reply:
There's a ton of conflicting information out there on this topic. Some people say the constructor should do the bare minimum, basically initialize fields. Others say it should do whatever work is necessary to get the object into a coherent state so other methods will return sensible values (not requiring another special method to be called to further initialize it before it's ready to be used). I think the design I'm hoping for is better thought of in terms of functional concepts.

With the design as is, the instance is not fully initialized when it is first constructed:
```py
po_file_generator = PoFileGenerator()
```
Notice it is possible, but makes no sense to reference `po_file_generator.generated_files` before calling
```py
po_file_generator.generate_translation_files(translations, {})
```
This does a second initialization. Before `generate_translation_files` is called the object is in a somewhat indeterminate state. Taking this further, how many times can or must I call `generate_translation_files`? Zero? One? What if I call it twice? If you move it into the constructor you don't have that problem because it's unconventional to call the constuctor any number of times other than once for a given instance.

Another way to handle this is to split your `PoFileGenerator` into two objects, so it can be used like this:
```py
po_file_generator = PoFileGenerator()
result = po_file_generator.generate_translation_files(translations, {})
try:
    for filename, filepath in result:
        # do stuff with filename/filepath
finally:
    result.cleanup()
```
In this case `PoFileGenerator` doesn't really need to be an object at all since it only has one method: it can just be a function (`generate_translation_files`) that returns the generated translation files. The `result` object is the more interesting part that will likely be a new object type (maybe call it `PoFiles`?) because it is both iterable and has a cleanup method. 

Does that clear things up?

Reply:
I also think that classes should receive all the information they need on initialization, so they're never in a nonsensical state.  I _definitely_ also think that functions should be preferred over objects, as they are the simplest possible logical unit - you provide inputs and get an output.  Of course there are instances where classes make more sense.

To your point about avoiding work in `__init__`, that also makes a lot of sense.  Daniel has some good suggestions, but I'll throw in another to specifically address that.  You can provide all of the inputs to the `__init__` method, but perform any non-trivial computation lazily.
```python
class PoFileGenerator(object):
    def __init__(self, translations, metadata):
        self.translations = translations
        self.metadata = metadata

    @cached_property
    def generated_files(self):
        # body of generate_translation_files
```
Depending on which of the options you chose, the bulk of the work occurs:

1. During `PoFileGenerator` initialization, which isn't great
1. As currently implemented, during `po_file_generator.generate_translation_files`, which is certainly obvious.  This has the downside of of `po_file_generator` being in a partially initialized state and later mutated.
1. In my example above, when `po_file_generator.generated_files` is first accessed.  I think this is better than in the constructor, but still arguably surprising.  You could use a memoized `generate_translation_files()` method instead of a property if you want to be clear that work is happening.
1. If this were implemented as a function, then when that function is called.  To my mind, this is the best option, as you provide all args up front, and it's clear that you expect work to be done.  You then pass around the results, rather than an object which maintains the results.  The limitation is of course is that sometimes you just need a class.

Reply:
@millerdev 
> Some people say the constructor should do the bare minimum

i think thats the group i belong to. i would still prefer keeping things simple.
tried [this](https://github.com/dimagi/commcare-hq/pull/21516/commits/acdf4f4812b58c9196f9ffe7c091af4cff145dcb) from the couple of suggestions you had. all looks okay but had to add [this](https://github.com/dimagi/commcare-hq/pull/21516/commits/acdf4f4812b58c9196f9ffe7c091af4cff145dcb#diff-389860842dade5566b99beb0b7021f2eR51) catch which is probably similar to what we are trying to resolve here.
Does this resolve the conflict?